### PR TITLE
Scalarization of savings and income shocks

### PR DESCRIPTION
**Pull request by viktoriakleinschmidt**
The function get_child_state_marginal_util_and_exp_max_values now works for single savings and income shocks instead of arrays exogenous_savings_grid and income_shock_draws.

### CODE REVIEW

File: src/dcegm/solve.py [hunk 1]
```diff
@@ -241,27 +241,41 @@ def backwards_induction(
...
+                        policy_array,
+                        value_array,
+                        compute_next_period_wealth,
+                        compute_marginal_utility,
+                        compute_value,
+                        get_state_specific_choice_set,
+                    )
+                    marginal_utility_weighted += marginal_util_weighted_shock
+                    max_exp_value_weighted += max_exp_value_weighted_shock
+                marginal_utilities[
```
Comment (MaxBlesch) [inline src/dcegm/solve.py]: Maybe just initialize the array with zeros and then add to each entry the calculated weighted marginal_utilities/max_exp_value

File: src/dcegm/egm.py [hunk 2]
```diff
@@ -151,10 +148,11 @@ def get_child_state_marginal_util_and_exp_max_value(
     child_node_choice_set = get_state_specific_choice_set(
```
Comment (MaxBlesch) [inline src/dcegm/egm.py]: The child specific choice set, the choice_values_child and choice_policies_child are independent of the savings and income shocks. Can we determine them outside of the for loops and then use them as function argument?

File: src/dcegm/egm.py [hunk 3]
```diff
@@ -343,23 +330,19 @@ def get_child_state_choice_specific_policy(
             (n_choices, n_quad_stochastic * n_grid_wealth).
 
     """
-    next_period_wealth_flat = next_period_wealth.flatten("F")
-
-    next_period_policy_interp = np.empty(
-        (child_node_choice_set.shape[0], next_period_wealth_flat.shape[0])
-    )
+    next_period_policy_interp = np.empty((child_node_choice_set.shape[0], 1))
```
Comment (MaxBlesch) [inline src/dcegm/egm.py]: The second dimension of the array is just 1 and therefore not needed. Just use a one dimensional array!

File: src/dcegm/egm.py [hunk 4]
```diff
@@ -380,25 +362,24 @@ def get_child_state_choice_specific_values(
...
+            period wealths. Shape (n_choices, n_quad_stochastic * n_grid_wealth).
 
     """
-    next_period_wealth_flat = next_period_wealth.flatten("F")
+
     next_period_value_interp = np.empty(
         (
             child_node_choice_set.shape[0],
-            next_period_wealth_flat.shape[0],
+            1,
```
Comment (MaxBlesch) [inline src/dcegm/egm.py]: Same here. Second dimension not needed.

**Review by MaxBlesch — CHANGES_REQUESTED** [review]
Very nice work so far. Just minor things!

**Comment by MaxBlesch** [comment]
Before we introduce jax vmap, we want to the user supplied functions to work on dictionaries instead of DataFrames. Also we want to have a check on the params DataFrame if it contains `taste_shock_scale` , `interest_rate` and `discount_factor`. These are all mandatory for now in the code.

File: src/dcegm/interpolation.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        endog_grid_child_state_choice[ind_low],
+        next_period_wealth,
+        compute_value,
+        compute_marginal_utility,
+        endog_grid_child_state_choice[1],
+        choice_values_child_state_choice[0],
+        choice,
+    )
+
+    return marg_utils, value_final
```
Comment (segsell) [inline src/dcegm/interpolation.py]: Why is the second output called ```value_final```?

File: src/dcegm/interpolation.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+    value_calc = compute_value(
+        consumption=new_wealth, next_period_value=value_min, choice=choice
+    )
+
+    constraint = new_wealth < endog_grid_min
+    value_final = constraint * value_calc + (1 - constraint) * value_interp
+
+    marg_utility_interp = compute_marginal_utility(policy_interp)
+
+    return marg_utility_interp, value_final
```
Comment (segsell) [inline src/dcegm/interpolation.py]: Same here. Why is it the final value?

**Review by MaxBlesch — APPROVED** [review]

File: src/dcegm/interpolation.py (hunk 6 above)
Comment (MaxBlesch) [reply src/dcegm/interpolation.py]: Renamed!

File: src/dcegm/interpolation.py (hunk 5 above)
Comment (MaxBlesch) [reply src/dcegm/interpolation.py]: Rename!

**Review by segsell — APPROVED** [review]
Great PR, good job everyone!!

**Pull request merged**